src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5875,310 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
+
+    ldout(s->cct, 10) << "s3-select query: failed to prase query; {" << s3select_syntax->get_error_description() << "}"<< dendl;
+
+    status = -1;
+  }
+  else
+  {
+    header_size = creare_header_records(m_buff_header);
+
+    m_result.append(m_buff_header,header_size); 
+
+    m_result.append(PAYLOAD_LINE);
+
+    status = m_s3_csv_object->run_s3select_on_stream(m_result,input,input_length,s->obj_size);
+    if(status<0)
+    {
+      m_result.append(m_s3_csv_object->get_error_description());
+    }
+  }
+
+  if (m_result.size() > strlen(PAYLOAD_LINE))
+  {
+    m_result.append(END_PAYLOAD_LINE);
+
+    int buff_len = create_message(m_result.data(), m_result.size() - 12, header_size);
+
+    s->formatter->write_bin_data(m_result.data(), buff_len);
+    if (op_ret < 0)
+      return op_ret;
+  }
+  rgw_flush_formatter_and_reset(s, s->formatter);
+
+  return status;
+}
+
+void RGWSelectObj_ObjStore_S3::convert_escape_seq(std::string & esc)
+{
+  if (esc.compare("\n") == 0)
+    esc = '\n';
+  else if (esc.compare("\t") == 0)
+    esc = '\t';
+  else if (esc.compare("\r") == 0)
+    esc = '\r';
+}
+
+int RGWSelectObj_ObjStore_S3::handle_aws_cli_parameters(std::string & sql_query)
+{
+
+  if(chunk_number !=0)
+    return 0;
+
+#define GT "&gt;"
+#define LT "&lt;"
+  if (m_s3select_query.find(GT) != std::string::npos)
+    boost::replace_all(m_s3select_query, GT, ">");
+  if (m_s3select_query.find(LT) != std::string::npos)
+    boost::replace_all(m_s3select_query, LT, "<");
+
+  //AWS cli s3select parameters
+  extract_by_tag("Expression", sql_query);
+
+  extract_by_tag("FieldDelimiter", m_column_delimiter);
+  convert_escape_seq(m_column_delimiter);
+
+  extract_by_tag("QuoteCharacter", m_quot);
+  convert_escape_seq(m_quot);
+
+  extract_by_tag("RecordDelimiter", m_row_delimiter);
+  convert_escape_seq(m_row_delimiter);
+  if (m_row_delimiter.size()==0) 
+    m_row_delimiter='\n';
+
+  extract_by_tag("QuoteEscapeCharacter", m_escape_char);
+  convert_escape_seq(m_escape_char);
+
+  extract_by_tag("CompressionType", m_compression_type);
+
+  if (m_compression_type.length()>0 && m_compression_type.compare("NONE") != 0)
+  {
+      ldout(s->cct, 10) << "RGW supports currently only NONE option for compression type" << dendl;
+      return -1;
+  }
+
+  extract_by_tag("FileHeaderInfo",m_header_info);
+
+  return 0;
+}
+
+int RGWSelectObj_ObjStore_S3::extract_by_tag(std::string tag_name,std::string & result)
+{
+  result = "";
+  size_t _qs = m_s3select_query.find("<" + tag_name + ">", 0) + tag_name.size() + 2;
+  if (_qs == std::string::npos)
+    return -1;
+  size_t _qe = m_s3select_query.find("</" + tag_name + ">", _qs);
+  if (_qe == std::string::npos)
+    return -1;
+
+  result = m_s3select_query.substr(_qs, _qe - _qs);
+
+  return 0;
+}
+
+int RGWSelectObj_ObjStore_S3::send_response_data(bufferlist &bl, off_t ofs, off_t len)
+{
+  if (len == 0)
+    return 0;
+
+  if (s->info.args.exists("select-type") && (chunk_number == 0))
```

Comment:
RGWOp has a virtual `get_params()` function. that's where we usually read the input and verify params. if we detect invalid input there, we can return an error to the client before `RGWGetObj::execute()` runs and starts reading object data